rotkehlchen/db/addressbook.py

```diff
@@ -36,7 +36,15 @@ def read_ctx(self, book_type: AddressbookType) -> Generator['DBCursor', None, No
             yield cursor
 
     @contextmanager
-    def write_ctx(self, book_type: AddressbookType) -> Generator['DBCursor', None, None]:
+    def write_ctx(
+            self,
+            book_type: AddressbookType,
+            given_cursor: Optional['DBCursor'] = None,
+    ) -> Generator['DBCursor', None, None]:
+        if given_cursor is not None:
```

Comment:
This makes no sense and is a hack. If you need to do that then you are not using the write_ctx correctly somewhere.

As for your comment:

> As for why to pass the cursor here -- it looked logical to me to give to write_ctx context manager all the data that is needed to determine which cursor to use.

Why does it look logical? It's not logical. It's hiding functionality and makes it absurdly easy to make mistakes. You should never have to pass a cursor here. Or in any other of the `write_cursor()` functions. Remove this entire thing.

You are corrupting a function with a very specific purpose by adding some logic which does not fit in there and is just to avoid adding logic in the proper place. The single caller where you need it. And thus introduce hidden fucntionality and if checks to every other caller.